### PR TITLE
Add support for channel name

### PR DESCRIPTION
**Pull request by rakshith91**
# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md##building-and-testing)
- [ ] Pull request includes test coverage for the included changes.

Fixes https://github.com/Azure/azure-sdk-for-python/issues/22519

### CODE REVIEW

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_helpers.py [hunk 1]
```diff
@@ -158,11 +158,14 @@ def _from_cncf_events(event):
         raise_with_traceback(ValueError, msg, err)
 
 
-def _build_request(endpoint, content_type, events):
+def _build_request(endpoint, content_type, events, channel_name=None):
```
Comment (lmazuel) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/_helpers.py:161]: Should be kwarg-only, even for an internal method

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py [hunk 2]
```diff
@@ -187,12 +187,14 @@ def send(self, events, **kwargs):
         :keyword str content_type: The type of content to be used to send the events.
          Has default value "application/json; charset=utf-8" for EventGridEvents,
          with "cloudevents-batch+json" for CloudEvents
+        :keyword str channel_name: Optional. Required only when publishing to partner namespaces with partner topic
```
Comment (lmazuel) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py:190]: We could debate if the main signature should be `send(self, events, * channel_name=None, **kwargs)`

cc @johanste

Comment (rakshith91) [reply sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py:190]: This feature will have very low usage (les than 1%) - i don't believe it's worth putting it in the method signature

File: sdk/eventgrid/azure-eventgrid/CHANGELOG.md [hunk 3]
```diff
@@ -1,11 +1,13 @@
 # Release History
 
-## 4.7.2 (Unreleased)
+## 4.8.0b1 (Unreleased)
```
Comment (rakshith91) [inline sdk/eventgrid/azure-eventgrid/CHANGELOG.md:3]:
```suggestion
## 4.9.0b1 (Unreleased)
```

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_version.py [hunk 4]
```diff
@@ -9,4 +9,4 @@
 # regenerated.
 # --------------------------------------------------------------------------
 
-VERSION = "4.7.2"
+VERSION = "4.8.0b1"
```
Comment (rakshith91) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/_version.py:12]:
```suggestion
VERSION = "4.9.0b1"
```

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py (hunk 2 above)
Comment (johanste) [reply sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py:190]: If we add the documentation, then we may as well make it keyword-only (as opposed to a `kwarg`). I'm still not entirely sure exactly who can make use of the feature. If it is truly 1p and a very limited number of 3p that will never grow that can make use of it, then adding it to the public surface area for everyone probably doesn't make sense. 

We pick one side of the fence. Document + keyword only, or don't document and kwarg.

Comment (rakshith91) [reply sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py:190]: "If it is truly 1p and a very limited number of 3p" - this is not entirely true. Anyone can become a partner and is not restricted to 1p.

There is absolutely no reason to not document it - let's do document  + keyword only

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py [hunk 5]
```diff
@@ -187,12 +192,14 @@ def send(self, events, **kwargs):
         :keyword str content_type: The type of content to be used to send the events.
          Has default value "application/json; charset=utf-8" for EventGridEvents,
          with "cloudevents-batch+json" for CloudEvents
+        :keyword str channel_name: Optional. Required only when publishing to partner namespaces with partner topic.
```
Comment (johanste) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py:195]: I would rephrase to avoid using the word "required" in the "Required only" disclaimer. Maybe something like "Used to specify the name of event channel when publishing to partner namespaces with (a?) partner topic"

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py [hunk 6]
```diff
@@ -187,12 +192,14 @@ def send(self, events, **kwargs):
...
+         The name of the event channel. For more details, visit
+         https://docs.microsoft.com/azure/event-grid/partner-events-overview
```
Comment (rakshith91) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py:197]:
```suggestion
        :keyword str channel_name: Optional.Used to specify the name of event channel when publishing to partner
         namespaces with partner topic. For more details, visit
         https://docs.microsoft.com/azure/event-grid/partner-events-overview
```

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/aio/_publisher_client_async.py [hunk 7]
```diff
@@ -187,12 +187,14 @@ async def send(self, events: SendType, **kwargs: Any) -> None:
         :keyword str content_type: The type of content to be used to send the events.
          Has default value "application/json; charset=utf-8" for EventGridEvents,
          with "cloudevents-batch+json" for CloudEvents
+        :keyword str channel_name: Optional. Required only when publishing to partner namespaces with partner topic.
+         The name of the event channel. For more details, visit
+         https://docs.microsoft.com/azure/event-grid/partner-events-overview
```
Comment (rakshith91) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/aio/_publisher_client_async.py:192]:
```suggestion
        :keyword str channel_name: Optional.Used to specify the name of event channel when publishing to partner
         namespaces with partner topic. For more details, visit
         https://docs.microsoft.com/azure/event-grid/partner-events-overview
```

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py [hunk 8]
```diff
@@ -124,8 +124,13 @@ def _policies(credential, **kwargs):
         return policies
 
     @distributed_trace
-    def send(self, events, **kwargs):
-        # type: (SendType, Any) -> None
+    def send(
+        self,
+        events: SendType,
+        *,
+        channel_name: str = None,
```
Comment (lmazuel) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py:131]: Isn't this a contradiction? If None is a value that is acceptable (acceptable being defined as doesn't raise), then I guess it should be `Optional[str]`? Does this code pass mypy? Do we have mypy enabled on EventGrid?

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py [hunk 9]
```diff
@@ -187,12 +192,14 @@ def send(self, events, **kwargs):
         :keyword str content_type: The type of content to be used to send the events.
          Has default value "application/json; charset=utf-8" for EventGridEvents,
          with "cloudevents-batch+json" for CloudEvents
+        :keyword str channel_name: Optional.Used to specify the name of event channel when publishing to partner
```
Comment (lmazuel) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py:195]:
```suggestion
        :keyword str channel_name: Optional. Used to specify the name of event channel when publishing to partner.
```

Comment (lmazuel) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py:195]: Is "partner" the name of the feature for EG team?

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/aio/_publisher_client_async.py [hunk 10]
```diff
@@ -187,12 +187,14 @@ async def send(self, events: SendType, **kwargs: Any) -> None:
         :keyword str content_type: The type of content to be used to send the events.
          Has default value "application/json; charset=utf-8" for EventGridEvents,
          with "cloudevents-batch+json" for CloudEvents
+        :keyword str channel_name: Optional.Used to specify the name of event channel when publishing to partner
```
Comment (lmazuel) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/aio/_publisher_client_async.py:190]:
```suggestion
        :keyword str channel_name: Optional. Used to specify the name of event channel when publishing to partner
```

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py (hunk 8 above)
Comment (rakshith91) [reply sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py:131]: it passes mypy - and we have it enabled - this is the precedence we have been using for keyword only (we don't tag it "optional")

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_helpers.py [hunk 11]
```diff
@@ -158,11 +158,15 @@ def _from_cncf_events(event):
         raise_with_traceback(ValueError, msg, err)
 
 
-def _build_request(endpoint, content_type, events):
+def _build_request(endpoint, content_type, events, **kwargs):
```
Comment (johanste) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/_helpers.py:161]: Nitpick: why not an optional keywords-only argument

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_helpers.py [hunk 12]
```diff
@@ -158,11 +158,15 @@ def _from_cncf_events(event):
...
+    channel_name = kwargs.pop('channel_name', None)
     serialize = Serializer()
     header_parameters = {}  # type: Dict[str, Any]
     header_parameters['Content-Type'] = serialize.header("content_type", content_type, 'str')
 
+    if channel_name:
+        header_parameters['aeg-channel-name'] = serialize.header("channel_name", channel_name, 'str')
```
Comment (johanste) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/_helpers.py:168]: What rules are there for escaping the channel name? I'm not sure why we need to do this in py3 in the first place for strings - wasn't this just for unicode/utf-8 handling?

Comment (rakshith91) [reply sdk/eventgrid/azure-eventgrid/azure/eventgrid/_helpers.py:168]: we wouldn't need it in python3 - this can go full py3 mode now that we are using keyword only instead of kwarg

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_helpers.py (hunk 11 above)
Comment (rakshith91) [reply sdk/eventgrid/azure-eventgrid/azure/eventgrid/_helpers.py:161]: changed it

File: sdk/eventgrid/azure-eventgrid/samples/sync_samples/sample_publish_to_channel.py [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+    CloudEvent(
+        type="Contoso.Items.ItemReceived",
+        source="/contoso/items",
+        data={
+            "itemSku": "Contoso Item SKU #1"
+        },
+        subject="Door1"
+    )
+],
+channel_name=channel_name)
```
Comment (lmazuel) [inline sdk/eventgrid/azure-eventgrid/samples/sync_samples/sample_publish_to_channel.py:42]: Indent don't feel right

File: sdk/eventgrid/azure-eventgrid/samples/async_samples/sample_publish_to_channel_async.py [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+            data={
+                "itemSku": "Contoso Item SKU #1"
+            },
+            subject="Door1"
+        )
+    ],
+    channel_name=channel_name)
+
+if __name__ == '__main__':
+    loop = asyncio.get_event_loop()
```
Comment (lmazuel) [inline sdk/eventgrid/azure-eventgrid/samples/async_samples/sample_publish_to_channel_async.py:46]: We should do asyncio.run:
https://github.com/Azure/azure-sdk-for-python/issues/21207

**Review by lmazuel — APPROVED** [review]
